## EDA - The Price of Art

In [31]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import kagglehub
import os
import re

In [32]:
# Download latest version
path = kagglehub.dataset_download("flkuhm/art-price-dataset")

print("Path to dataset files:", path)

Path to dataset files: /home/codespace/.cache/kagglehub/datasets/flkuhm/art-price-dataset/versions/1


In [33]:
# 1. First, let's see what files are inside the directory we just downloaded.
print("Files in the directory:", os.listdir(path))

Files in the directory: ['artDataset.csv', 'artDataset']


In [34]:
# 2. Now, we need the exact filename. Let's assume from the Kaggle page that the file is named 'artDataset'
filename = 'artDataset.csv'

In [35]:
# 3. We create the full, correct path to the file. 
# os.path.join is the best way to do this, as it works on any operating system (Windows, Mac, Linux).
full_path = os.path.join(path, filename)

In [36]:
# 4. Finally, we read the CSV file into our pandas DataFrame.
try:
    df = pd.read_csv(full_path)
    print("\nDataset loaded successfully!")
except FileNotFoundError:
    print(f"\nError: The file '{filename}' was not found in the directory '{path}'.")
    print("Please check the filename from the 'os.listdir(path)' output and update the 'filename' variable.")



Dataset loaded successfully!


In [37]:
df = pd.read_csv(full_path)
df

,Unnamed: 0,price,artist,title,yearCreation,signed,condition,period,movement
0,0,28.500 USD,Tommaso Ottieri,Bayreuth Opera,2021,Signed on verso,This work is in excellent condition.,Contemporary,Baroque
1,1,3.000 USD,Pavel Tchelitchew,Drawings of the Opera,First Half 20th Century,Signed and titled,Not examined out of frame.No obvious signs of ...,Post-War,Surrealism
2,2,5.000 USD,Leo Gabin,Two on Sidewalk,2016,"Signed, titled and dated on verso",This work is in excellent condition.,Contemporary,Abstract
3,3,5.000 USD,Matthias Dornfeld,Blumenszene,2010,"Signed, titled and dated on the reverse with t...",This work is in excellent condition.There is m...,Contemporary,Abstract
4,4,2.500 USD,Alexis Marguerite Teplin,Feverish Embarkation,2001,Signed on verso,This work is in excellent condition.,Contemporary,Abstract
...,...,...,...,...,...,...,...,...,...
749,749,680 USD,Jane Kent,Miracle Grow #17,2012,Signed and dated on lower right.,Not examined out of frame.No obvious signs of ...,Contemporary,Abstract
750,750,1.275 USD,Gary Bower,Rolph Series,1970,[nan],Not examined out of frame.Significant undulati...,Contemporary,Geometric Abstraction
751,751,680 USD,Jane Kent,Untitled,2012,[nan],Not examined out of frame.No apparent imperfec...,Contemporary,Geometric Abstraction
752,752,1.275 USD,T. L. Solien,Juniper,1986,[nan],Not examined outside of frame.Pinholes at edge...,Contemporary,Abstract


In [38]:
# Let's understand the dimensions and the columns of our dataset.
# This gives us a quick overview without printing thousands of rows.

print(f"The DataFrame has {df.shape[0]} rows and {df.shape[1]} columns.")
print("\nColumn Names:")
print(df.columns.tolist())

The DataFrame has 754 rows and 9 columns.

Column Names:
['Unnamed: 0', 'price', 'artist', 'title', 'yearCreation', 'signed', 'condition', 'period', 'movement']


In [39]:
# The 'Unnamed: 0' column is likely an old index from the CSV file.
# It's redundant, so let's drop it to clean our DataFrame.

if 'Unnamed: 0' in df.columns:
    df.drop('Unnamed: 0', axis=1, inplace=True)
    print("Dropped the 'Unnamed: 0' column.")
else:
    print("'Unnamed: 0' column not found. No action taken.")

# Let's verify the change
print("\nUpdated Column Names:")
print(df.columns.tolist())

Dropped the 'Unnamed: 0' column.

Updated Column Names:
['price', 'artist', 'title', 'yearCreation', 'signed', 'condition', 'period', 'movement']


In [40]:
# Get a concise summary of the DataFrame to check data types and missing values.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 754 entries, 0 to 753
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   price         754 non-null    str  
 1   artist        753 non-null    str  
 2   title         754 non-null    str  
 3   yearCreation  754 non-null    str  
 4   signed        754 non-null    str  
 5   condition     754 non-null    str  
 6   period        754 non-null    str  
 7   movement      754 non-null    str  
dtypes: str(8)
memory usage: 47.3 KB


In [41]:
# Get descriptive statistics for the numerical columns.
df.describe().T

,count,unique,top,freq
price,754,108,800 USD,124
artist,753,454,Russell Young,17
title,754,679,Untitled,35
yearCreation,754,136,2012,34
signed,754,390,[nan],153
condition,754,376,Excellent condition.,82
period,754,5,Contemporary,414
movement,754,34,Realism,177


## Initial Findings from Data Inspection
After the initial data loading and inspection using .info() and .describe(), several key characteristics and potential issues have been identified in the dataset:

- Data Type Mismatch: All columns, including critical ones like price and yearCreation, are currently stored as objects (str). This prevents any numerical or time-series analysis.
- Price Column: The price column contains the ' USD' suffix, which needs to be removed before conversion to a numeric type. The most frequent price is '$800', which may indicate a specific category of art or a default price that requires further investigation.
Missing Values:
- The artist column has one official missing value (NaN). Given that there is only one, this warrants a specific investigation of the corresponding title to make an informed imputation decision (e.g., is it a famous work, or by an 'Anonymous' artist?).
- The signed column contains the literal string '[nan]' as its most frequent value. This is a hidden missing value that needs to be converted to a proper NaN for accurate analysis.
- High Cardinality: The condition column has 376 unique values, which is too many for effective analysis. It will likely require grouping or simplification later in the project.

## Proposed Cleaning & Preprocessing Plan
To prepare the dataset for exploratory analysis and visualization, the following steps will be taken in order:

- Investigate Missing Artist: Before any bulk transformations, we will isolate the single row with the missing artist value. By examining its title and other attributes, we will decide on the most appropriate value to impute (e.g., "Anonymous", "Unknown", or the actual artist's name if identifiable).
- Convert price to Numeric:
Remove the ' USD' suffix and any commas from the price column.
Convert the column to a numeric (float) data type.
- Convert yearCreation to Numeric:
Inspect the column for non-numeric characters.
Convert the column to a numeric (integer) data type.
- Handle Hidden Missing Values:
Convert the literal '[nan]' strings in the signed column to actual NaN values.
Decide on a strategy for the now-missing values in the signed column (e.g., imputation with a placeholder like 'Not Signed').
- (Future) Simplify Categorical Columns:
Once the main structural issues are resolved, we will address the high cardinality in the condition column to make it more analyzable.

## STEP I - Investigating the single missing 'artist' value:

In [42]:
# Let's find the row where the 'artist' is missing (NaN)
missing_artist_row = df[df['artist'].isnull()]

# Display the full information for this row
print("Investigating the row with the missing artist:")
print(missing_artist_row.T) # Using .T to transpose it for better readability

Investigating the row with the missing artist:
                                                            725
price                                                 1.275 USD
artist                                                      NaN
title                                                     [nan]
yearCreation                                              [nan]
signed                      Signed and dated in pencil to verso
condition     Not examined out of frame.Minor sheet undulati...
period                                             Contemporary
movement                                                Realism


In [43]:
# Based on our investigation of row 725:
# We will impute missing values for 'artist' and 'title' to avoid data loss, as the row contains valid price and category information.

df.loc[df['artist'].isnull(), 'artist'] = 'Unknown Artist'
df.loc[df['title'] == '[nan]', 'title'] = 'Untitled / Unknown'

# Verify the fix
print(df.loc[725, ['artist', 'title', 'price']])

artist        Unknown Artist
title     Untitled / Unknown
price              1.275 USD
Name: 725, dtype: str


## Handling Specific Missing Values
During the initial inspection, a specific case (index 725) was identified where the 'artist' and 'title' were missing, yet the record contained valid transaction data (price) and stylistic metadata (Contemporary Realism).

Instead of removing this entry—which would bias our price analysis—I decided to impute these missing categorical fields with 'Unknown Artist' and 'Untitled / Unknown'. This preserves the financial data for the Power BI dashboard while maintaining the integrity of the categorical filters.

## STEP II - Price Sanitization

In [44]:
# We create a copy to avoid SettingWithCopyWarning and keep the original for reference if needed
df['price'] = df['price'].str.replace(' USD', '', regex=False) # Remove suffix
df['price'] = df['price'].str.replace('.', '', regex=False)    # Remove thousand separator (.)
df['price'] = df['price'].astype(float)                       # Convert to numeric

# Let's check the statistics to make sure the prices make sense now
print("Price Statistics:")
print(df['price'].describe())

Price Statistics:
count       754.000000
mean       6760.769231
std       12828.990543
min         595.000000
25%        1000.000000
50%        2500.000000
75%        7000.000000
max      160000.000000
Name: price, dtype: float64


In [45]:
print(df['price'])

0      28500.0
1       3000.0
2       5000.0
3       5000.0
4       2500.0
        ...   
749      680.0
750     1275.0
751      680.0
752     1275.0
753      680.0
Name: price, Length: 754, dtype: float64


## Data Normalization: Price Column:
The price column was originally loaded as a string object, containing currency suffixes ('USD') and period-based thousand separators (e.g., '1.275').

To enable quantitative analysis and ensure compatibility with Power BI's numerical schemas, I performed a sanitization process:

1.- Stripped the 'USD' currency descriptor.

2.- Removed the period separators to prevent Python from interpreting them as decimal points (as the dataset follows a format where 1.275 equals 1,275).

3.- Cast the resulting strings into a float data type.

This transformation allows for the calculation of total market value, average pricing per movement, and other key financial metrics."

In [46]:
# Convert to integer since we don't have decimals
df['price'] = df['price'].astype(int)

# Now let's check the head to see them nice and clean
print(df['price'].head())

0    28500
1     3000
2     5000
3     5000
4     2500
Name: price, dtype: int64


Refining Price Data Type
After the initial cleaning of the price column, a statistical review revealed that all transaction values in the dataset were whole numbers (integers), with no decimal precision required.

To improve data readability and optimize memory usage, I converted the price column from float to int. This decision ensures that the EDA reflects the actual nature of the art market entries in this dataset and simplifies the visual representation in the final Power BI dashboard.

## STEP III - GLOBAL NULL STANDARDIZATION

In [47]:
# We replace the literal string '[nan]' with the official NumPy representation of Null
df.replace('[nan]', np.nan, inplace=True)

,price,artist,title,yearCreation,signed,condition,period,movement
0,28500,Tommaso Ottieri,Bayreuth Opera,2021,Signed on verso,This work is in excellent condition.,Contemporary,Baroque
1,3000,Pavel Tchelitchew,Drawings of the Opera,First Half 20th Century,Signed and titled,Not examined out of frame.No obvious signs of ...,Post-War,Surrealism
2,5000,Leo Gabin,Two on Sidewalk,2016,"Signed, titled and dated on verso",This work is in excellent condition.,Contemporary,Abstract
3,5000,Matthias Dornfeld,Blumenszene,2010,"Signed, titled and dated on the reverse with t...",This work is in excellent condition.There is m...,Contemporary,Abstract
4,2500,Alexis Marguerite Teplin,Feverish Embarkation,2001,Signed on verso,This work is in excellent condition.,Contemporary,Abstract
...,...,...,...,...,...,...,...,...
749,680,Jane Kent,Miracle Grow #17,2012,Signed and dated on lower right.,Not examined out of frame.No obvious signs of ...,Contemporary,Abstract
750,1275,Gary Bower,Rolph Series,1970,NaN,Not examined out of frame.Significant undulati...,Contemporary,Geometric Abstraction
751,680,Jane Kent,Untitled,2012,NaN,Not examined out of frame.No apparent imperfec...,Contemporary,Geometric Abstraction
752,1275,T. L. Solien,Juniper,1986,NaN,Not examined outside of frame.Pinholes at edge...,Contemporary,Abstract


In [48]:
# Let's verify how many ACTUAL missing values we have now in each column
print("Updated Missing Values Count:")
print(df.isnull().sum())

Updated Missing Values Count:
price             0
artist            0
title             0
yearCreation     19
signed          153
condition         7
period            1
movement          4
dtype: int64


In [49]:
# Let's see specifically the columns we were worried about
print("\nCheck for specific columns:")
print(df[['yearCreation', 'signed', 'condition']].head(10))


Check for specific columns:
               yearCreation  \
0                      2021   
1  First Half 20th Century    
2                      2016   
3                      2010   
4                      2001   
5                      1977   
6                      2007   
7                      1993   
8                      2014   
9                      1993   

                                              signed  \
0                                    Signed on verso   
1                                  Signed and titled   
2                  Signed, titled and dated on verso   
3  Signed, titled and dated on the reverse with t...   
4                                    Signed on verso   
5  Signed lower right recto; numbered lower left ...   
6        Numbered and signed on bottom corner, recto   
7  Signed lower right recto; numbered lower left ...   
8  Signed lower right recto; numbered lower left ...   
9  Signed lower right recto; numbered lower left ...   

            

# Data Quality Audit & Initial Findings
Upon standardizing the dataset and converting pseudo-null strings ([nan]) into actual null objects, a clear picture of the data quality emerged:

1.- Year of Creation (yearCreation): Only 19 records are completely missing. However, the column contains a mix of specific years (e.g., '2021') and categorical periods (e.g., 'First Half 20th Century'), requiring a custom parsing strategy.

2.- Authentication (signed): This is the most incomplete field with 153 missing values. This suggests that while price and artist are well-documented, the physical details of the signature are often omitted in the catalog.

3.- Physical State (condition): With only 7 missing values, this column is remarkably complete, though it exhibits high textual complexity that will require grouping.

4.- Artistic Context (period & movement): These fields are nearly perfect, with only 1 and 4 missing values respectively, providing a solid foundation for categorical analysis.

# Strategic Decision: Data Recovery over Deletion
Instead of dropping rows with missing or descriptive temporal data, I have opted for a Data Recovery approach:

1.- Imputation by Context: For the 19 missing years, I will leverage the movement and period columns to estimate the era.

2.- Mapping Categorical Periods: Descriptions like 'First Half 20th Century' will be mapped to their chronological mid-points (e.g., 1925) to maintain a continuous timeline for Power BI visualizations.

3.- Labeling: Missing values in the signed column will be labeled as 'Not Specified' to serve as a valid filter category rather than a 'Blank' entry.

## STEP IV - ADVANCED YEAR RECOVERY & CONDITION GROUPING 

The yearCreation column contained a mixture of specific dates and broad period descriptors. To enable time-series analysis, I implemented a custom recovery function.

Following a conservative estimation strategy, records labeled as 'First Half' of a century were mapped to the year '30' (e.g., 1830), and 'Second Half' records were mapped to the year '80' (e.g., 1880). This provides a cleaner chronological distribution for the upcoming dashboard while preserving 4-digit year formats.

In [50]:
def clean_year_logic(year_value):
    """
    Standardizes year values. Handles numeric types, specific years in strings,
    and estimates years for historical periods like 'First/Second Half'.
    """
    # 1. Handle existing nulls
    if pd.isna(year_value):
        return np.nan
    
    # 2. If it's already a number (int/float), return as int
    if isinstance(year_value, (int, float)):
        return int(year_value)
    
    # 3. Process string values
    year_str = str(year_value).strip().lower()
    
    # Logic for 'First Half' of the century (e.g., 1830)
    if 'first half' in year_str:
        century_match = re.search(r'(\d{2})', year_str)
        if century_match:
            century = int(century_match.group(1)) - 1
            return int(f"{century}30")
            
    # Logic for 'Second Half' of the century (e.g., 1880)
    if 'second half' in year_str:
        century_match = re.search(r'(\d{2})', year_str)
        if century_match:
            century = int(century_match.group(1)) - 1
            return int(f"{century}80")
            
    # Extract any 4-digit number (covers "c. 1920", "2021.", etc.)
    match = re.search(r'\d{4}', year_str)
    if match:
        return int(match.group())
    
    return np.nan

# Apply the refined function
df['year_numeric'] = df['yearCreation'].apply(clean_year_logic)

# Final validation
print(f"Original Nulls: {df['yearCreation'].isnull().sum()}")
print(f"Final Nulls after refined recovery: {df['year_numeric'].isnull().sum()}")

Original Nulls: 19
Final Nulls after refined recovery: 47


In [51]:
# Investigating the discrepancy
# We want to see rows where 'yearCreation' had a value but 'year_numeric' is now NaN
discrepancy = df[df['yearCreation'].notna() & df['year_numeric'].isna()]

print(f"Number of problematic rows: {len(discrepancy)}")
print("\nSample of problematic 'yearCreation' values:")
print(discrepancy['yearCreation'].unique())

# Also, let's check the data types of the original column
print("\nData type of yearCreation column:")
print(df['yearCreation'].dtype)

Number of problematic rows: 28

Sample of problematic 'yearCreation' values:
<StringArray>
[                                                                                                                    'Mid 20th Century ',
                                                                                                                         '19th Century ',
 '3D printed using ABS, PLA plastics, resin, automobile paints, etched brass, dry transfers, acrylic mirror, batik fabric, quartz clock',
                                                                                                                    'Late 20th Century ',
                                                                                                                    'Late 19th Century ',
                                                                                                                         '20th Century ',
                                                                                 

During the initial cleaning, I noticed the null count increased from 19 to 47. Upon investigation, I discovered that the dataset contained artistic descriptors instead of dates (e.g., 'Late 20th Century' or technical descriptions of 3D printing).

To preserve data integrity and minimize loss, I implemented a granular mapping strategy:

- Early Century: Mapped to year '10'.
- Mid Century: Mapped to year '50'.
- Late Century: Mapped to year '90'.
- 3D Printed Artifacts: Assigned to 2010, marking the era of ABS/PLA popularization in contemporary art.

This iterative process reduced the data loss back to the original 19 missing values, ensuring a more robust dataset for the Power BI timeline

In [52]:
def clean_year_logic_final(year_value):
    if pd.isna(year_value):
        return np.nan
    
    year_str = str(year_value).strip()
    year_lower = year_str.lower()
    
    # 1. Broad Century Logic with your custom precision:
    # Early=10, First Half=30, Mid=50, Second Half=80, Late=90
    century_match = re.search(r'(\d{2})', year_lower)
    if century_match and 'century' in year_lower:
        century_prefix = int(century_match.group(1)) - 1
        if 'early' in year_lower:
            return int(f"{century_prefix}10")
        if 'first half' in year_lower:
            return int(f"{century_prefix}30")
        if 'mid' in year_lower:
            return int(f"{century_prefix}50")
        if 'second half' in year_lower:
            return int(f"{century_prefix}80")
        if 'late' in year_lower:
            return int(f"{century_prefix}90")
        # Default for century only
        return int(f"{century_prefix}50")

    # 2. Saving the "3D Printed" outlier (ABS/PLA popularization c. 2010)
    if '3d printed' in year_lower or 'abs' in year_lower:
        return 2010

    # 3. Extract standard 4-digit years
    match = re.search(r'\d{4}', year_lower)
    if match:
        return int(match.group())
    
    return np.nan

# Applying the ultimate logic
df['year_numeric'] = df['yearCreation'].apply(clean_year_logic_final)
print(f"Original Nulls: {df['yearCreation'].isnull().sum()}")
print(f"Final Nulls after investigation: {df['year_numeric'].isnull().sum()}")

Original Nulls: 19
Final Nulls after investigation: 19


# STEP IV.I. - Deep Dive: Investigating Missing Temporal Data through Artistic Periods

After the initial cleaning, 19 records still lack a specific creation year. To avoid data deletion, I will now investigate the period attribute of these specific rows.

By analyzing the distribution of artistic periods within the entire dataset, I can establish a historically accurate timeframe for each missing record. The following step performs a frequency count of all periods to identify the chronological context we are working with

In [53]:
# 1. First, let's see all available periods in the entire dataset
print("--- Global Period Distribution ---")
print(df['period'].value_counts())

--- Global Period Distribution ---
period
Contemporary    414
Post-War        285
Modern           42
19th Century     12
Name: count, dtype: int64


In [ ]:
# 2. Now, let's isolate the 19 rows that still have a null 'year_numeric' and see what period they belong to
missing_years_filter = df['year_numeric'].isna()
print("\n--- Periods associated with the 19 missing years ---")
print(df.loc[missing_years_filter, 'period'].value_counts())


--- Periods associated with the 19 missing years ---
period
Post-War        10
Contemporary     7
Modern           2
Name: count, dtype: int64


Based on the frequency distribution, the 19 missing records belong exclusively to the Post-War (10), Contemporary (7), and Modern (2) periods.

To fill these gaps, I am implementing a targeted imputation using historical benchmarks for each category:

- Post-War: Assigned to 1960 (the height of the post-war artistic boom).
- Modern: Assigned to 1920 (a core year for Modernism). 
- Contemporary: Assigned to 2000 (a neutral midpoint for recent works).

This approach ensures that the data remains statistically viable for time-series analysis while honoring the historical context of each piece.

In [56]:
# Defining the historical mapping for the missing periods
period_imputation_map = {
    'Post-War': 1960,
    'Contemporary': 2000,
    'Modern': 1920
}

# Applying the imputation specifically to the nulls in year_numeric
df['year_numeric'] = df.apply(
    lambda row: period_imputation_map.get(row['period'], row['year_numeric']) 
    if pd.isna(row['year_numeric']) else row['year_numeric'],
    axis=1
)

# Converting the column to integer now that it's fully populated
df['year_numeric'] = df['year_numeric'].astype(int)

# Final verification
print(f"Total Nulls in year_numeric after strategic imputation: {df['year_numeric'].isnull().sum()}")

Total Nulls in year_numeric after strategic imputation: 0


The chronological recovery phase is now complete. By combining RegEx-based extraction, forensic string analysis (handling 'Early/Mid/Late' descriptors), and domain-driven imputation based on artistic periods, I have successfully eliminated all missing values in the year_numeric column.

The dataset now has a 100% completion rate for its temporal data, transitioning from 19 original nulls and several ambiguous strings to a clean, integer-based format ready for time-series visualization in Power BI.

## STEP V - SIMPLIFYING CONDITION (For Power BI Visualization)

The original condition field presented high cardinality with over 370 unique string descriptions, which is unsuitable for effective dashboard filtering.

I performed keyword extraction to consolidate these descriptions into four standardized categories: 'Excellent', 'Good', 'Fair/Aged', and 'Other'. This transformation simplifies the data structure and prepares the field for categorical comparison in Power BI.

In [57]:
def group_condition_categories(text):
    if pd.isna(text):
        return 'Not Specified'
    
    text = text.lower()
    
    # Keyword extraction for grouping
    if 'excellent' in text:
        return 'Excellent'
    if 'very good' in text or 'good' in text:
        return 'Good'
    if 'fair' in text or 'worn' in text:
        return 'Fair/Aged'
    
    return 'Other'

# Applying the grouping logic
df['condition_grouped'] = df['condition'].apply(group_condition_categories)

# Visualizing the reduction in cardinality
print(f"Original unique conditions: {df['condition'].nunique()}")
print(f"Grouped conditions: {df['condition_grouped'].unique()}")

Original unique conditions: 375
Grouped conditions: <StringArray>
['Excellent', 'Other', 'Good', 'Not Specified']
Length: 4, dtype: str


The condition field originally presented a high cardinality of 375 unique strings, ranging from detailed technical reports to brief notes. Such fragmentation is counterproductive for data visualization and filtering.

To optimize this for Business Intelligence, I implemented a keyword-based classification logic to consolidate these records into four actionable categories:

- Excellent: Works in top-tier preservation state.
- Good: Works with minor signs of age or standard wear.
- Not Specified: Missing or ambiguous data.
- Other: Specialized or complex condition reports.

This transformation reduces noise and allows for meaningful categorical analysis of how an artwork's state affects its market value

In [58]:
def group_condition_logic(text):
    if pd.isna(text) or text == '':
        return 'Not Specified'
    
    text = str(text).lower()
    
    # Logic for grouping
    if 'excellent' in text:
        return 'Excellent'
    if 'good' in text or 'very good' in text:
        return 'Good'
    if 'not specified' in text or 'unknown' in text:
        return 'Not Specified'
    
    # Everything else goes to 'Other'
    return 'Other'

# Applying the transformation
df['condition_grouped'] = df['condition'].apply(group_condition_logic)

# Verification of the reduction
print(f"Original unique conditions: {df['condition'].nunique()}")
print(f"New grouped categories: {df['condition_grouped'].unique()}")
print("\nDistribution per category:")
print(df['condition_grouped'].value_counts())

Original unique conditions: 375
New grouped categories: <StringArray>
['Excellent', 'Other', 'Good', 'Not Specified']
Length: 4, dtype: str

Distribution per category:
condition_grouped
Good             300
Excellent        269
Other            178
Not Specified      7
Name: count, dtype: int64


## SECOND GLANCE:

Following the initial cleaning phase, the dataset reflects a 99% completion rate. To achieve total data integrity, a targeted audit is performed on the remaining missing values within the period and movement attributes.

In [59]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 754 entries, 0 to 753
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   price              754 non-null    int64
 1   artist             754 non-null    str  
 2   title              754 non-null    str  
 3   yearCreation       735 non-null    str  
 4   signed             601 non-null    str  
 5   condition          747 non-null    str  
 6   period             753 non-null    str  
 7   movement           750 non-null    str  
 8   year_numeric       754 non-null    int64
 9   condition_grouped  754 non-null    str  
dtypes: int64(2), str(8)
memory usage: 59.0 KB


In [60]:
df.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
price,754.0,NaN,NaN,NaN,6760.769231,12828.990543,595.0,1000.0,2500.0,7000.0,160000.0
artist,754,455,Russell Young,17,NaN,NaN,NaN,NaN,NaN,NaN,NaN
title,754,679,Untitled,35,NaN,NaN,NaN,NaN,NaN,NaN,NaN
yearCreation,735,135,2012,34,NaN,NaN,NaN,NaN,NaN,NaN,NaN
signed,601,389,Signed lower right,29,NaN,NaN,NaN,NaN,NaN,NaN,NaN
condition,747,375,Excellent condition.,82,NaN,NaN,NaN,NaN,NaN,NaN,NaN
period,753,4,Contemporary,414,NaN,NaN,NaN,NaN,NaN,NaN,NaN
movement,750,33,Realism,177,NaN,NaN,NaN,NaN,NaN,NaN,NaN
year_numeric,754.0,NaN,NaN,NaN,1989.419098,26.284872,1850.0,1978.0,1990.0,2008.0,2050.0
condition_grouped,754,4,Good,300,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## Foundings on the Second Glance:

Upon review of the initial cleaning results, a high completion rate is observed. However, a specific discrepancy in categorical density remains. The objective of this phase is to transition from a 'mostly clean' dataset to a state of 100% data integrity through targeted intervention.

Audit Observations:

- Column period: A single missing value (1 record) is identified.

- Column movement: Four missing values (4 records) are identified.

- Column signed: Significant fragmentation exists with over 380 unique strings and 153 missing entries.

Action Plan:

Redundancy Cleanup: The yearCreation column is removed, as it has been superseded by the validated year_numeric attribute.

Surgical Imputation: Manual research is performed for the 5 missing categorical records to ensure historical accuracy without distorting the dataset with generic placeholders.

Boolean Transformation: The signed column is standardized into a binary format (Yes/No) to facilitate correlation analysis between artist signatures and market pricing

## STEP VI - REMOVAL OF REDUNDANT FEATURES:
Once the year_numeric column is validated and fully populated, the original yearCreation attribute becomes redundant. It is removed to streamline the dataset and prevent confusion during the dashboard development phase.

In [61]:
# Dropping the original yearCreation column
df.drop(columns=['yearCreation'], inplace=True)

In [62]:
# Verification
print("Column 'yearCreation' removed. Current columns:")
print(df.columns.tolist())

Column 'yearCreation' removed. Current columns:
['price', 'artist', 'title', 'signed', 'condition', 'period', 'movement', 'year_numeric', 'condition_grouped']


## STEP VII - IMPUTATION OF PERIOD AND MOVEMENT
A targeted identification of the remaining null values in period (1) and movement (4) is performed. These records are isolated to be manually categorized, ensuring 100% data density without relying on automated generic fillers.

In [63]:
# Identifying the specific rows to fix
lone_period = df[df['period'].isna()]
fantastic_four_movement = df[df['movement'].isna()]

print("TARGETED INSPECTION: MISSING PERIOD")
display(lone_period[['artist', 'title', 'year_numeric']])

print("\n TARGETED INSPECTION: MISSING MOVEMENTS")
display(fantastic_four_movement[['artist', 'title', 'year_numeric']])

TARGETED INSPECTION: MISSING PERIOD


,artist,title,year_numeric
543,Ella Delyanis,"Study, Storm Over Bell Pond",1983



 TARGETED INSPECTION: MISSING MOVEMENTS


,artist,title,year_numeric
12,Zhang Yu,96.6.2,2000
343,CARL VAN VECHTEN,Dizzy Gillespie,1955
543,Ella Delyanis,"Study, Storm Over Bell Pond",1983
570,Richard Ballard,Cypress,1990


Following a targeted investigation, the remaining missing values in the period and movement columns are manually assigned. This ensures historical accuracy and maintains a 100% completion rate for categorical analysis.

- Ella Delyanis: Assigned to 'Contemporary' and 'Realism'.

- Zhang Yu: Assigned to 'Contemporary'.

- Carl Van Vechten: Assigned to 'Modernism'.

- Richard Ballard: Assigned to 'Abstract Art'.

In [64]:
# Filling the 'Period' null (Index 543)
df.at[543, 'period'] = 'Contemporary'

# Filling the 'Movement' nulls
df.at[12, 'movement'] = 'Contemporary'          # Zhang Yu
df.at[343, 'movement'] = 'Modernism'            # Carl Van Vechten
df.at[543, 'movement'] = 'Realism'              # Ella Delyanis
df.at[570, 'movement'] = 'Abstract Art'         # Richard Ballard

# Final check for Period and Movement
print(f"Nulls in Period: {df['period'].isnull().sum()}")
print(f"Nulls in Movement: {df['movement'].isnull().sum()}")

Nulls in Period: 0
Nulls in Movement: 0


## STEP VIII - Signature Standardization: Boolean Logic Implementation

The 'signed' column currently contains highly fragmented descriptions (e.g., 'Signed lower right', 'Stamped', 'Initialed') and several missing values. To enable quantitative analysis and effective filtering in Power BI, this field is transformed into a binary format (Yes/No).

Standardization Logic:

- Yes: Any record containing keywords such as 'signed', 'stamped', 'monogrammed', or 'initialed'.

- No: Records explicitly stated as 'unsigned' or missing (NaN). In the professional art market, the absence of a signature description typically implies an unsigned work.

# STEP VIII.I - Signature Categorical Exhaustion

In [65]:
# Extracting every unique signature string to build a robust Boolean logic
unique_signatures = df['signed'].unique()

print(f"--- TOTAL UNIQUE SIGNATURE DESCRIPTIONS: {len(unique_signatures)} ---")
for sig in sorted([str(x) for x in unique_signatures]):
    print(f"- {sig}")

print(f"\nReminder: There are {df['signed'].isnull().sum()} null records (NaN).")

--- TOTAL UNIQUE SIGNATURE DESCRIPTIONS: 390 ---
- A Flip Schulke, West Palm Beach studio label on the reverse, annotated ‘1964’ in red ink.
- A Galerie Walter Storms postcard on the reverse of the mount, with credit, title and date engraved on the side.
- All three photographs signed and numbered verso
- Apel inscribed at lower right recto
- Artist and printer chops on right front.
- Artist name inscribed verso
- Artist signature and date incorporated in work
- Artist signature and date lower left recto
- Artist signature and date lower right recto
- Artist signature and edition in pencil lower right recto.
- Artist signature bottom right recto
- Artist signature lower right recto
- Artist signature lower right.
- Artist signature on back.
- Artist signed in pen lower right
- Barbara Schwartz 2001 in pencil on bottom right recto.
- Blindstamped by the photographer’s archive in the margin.Signed and copyrighted in pencil on the reverse, a Flip Schulke, West Palm Beach studio label on t

# STEP VIII.II - Signature Standardization: Boolean Logic Implementation

The signed attribute is converted from 390 heterogeneous strings into a binary Boolean (Yes/No) format.

Decision Matrix:

- Yes: Assigned to any record containing evidence of authorship, including explicit keywords (signed, stamped, initialed, inscribed, signature) or specific archival labels and markings.

- No: Assigned to missing values (NaN) and records explicitly marked as 'unsigned'. In the fine art trade, the absence of documented signature data is a standard proxy for an unsigned work.

This transformation creates a high-impact feature for price correlation analysis.

In [66]:
def transform_signature(text):
    # If it's null (NaN), it's a 'No'
    if pd.isna(text) or str(text).lower() == 'nan':
        return 'No'
    
    # Convert to string and clean spaces
    text_clean = str(text).lower().strip()
    
    # Check for explicit 'unsigned' or 'not signed'
    if 'unsigned' in text_clean or 'not signed' in text_clean:
        return 'No'
    
    # If the text is empty after cleaning
    if not text_clean:
        return 'No'
    
    # Everything else in this specific dataset (labels, stamps, names) counts as 'Yes'
    return 'Yes'

# Applying the logic
df['is_signed'] = df['signed'].apply(transform_signature)

# Final check of the distribution
print("--- SIGNATURE STATUS DISTRIBUTION ---")
print(df['is_signed'].value_counts())

# Safety check: show some 'No' examples to make sure they are correct
print("\n--- SAMPLE OF 'NO' RECORDS ---")
display(df[df['is_signed'] == 'No'][['signed', 'is_signed']].head(10))

--- SIGNATURE STATUS DISTRIBUTION ---
is_signed
Yes    601
No     153
Name: count, dtype: int64

--- SAMPLE OF 'NO' RECORDS ---


,signed,is_signed
32,NaN,No
75,NaN,No
77,NaN,No
114,NaN,No
115,NaN,No
117,NaN,No
120,NaN,No
122,NaN,No
123,NaN,No
124,NaN,No


## STEP IX - QUALITY CHECK:

Before the final export, a comprehensive validation is performed to ensure zero data loss and perfect attribute density. This step confirms that every cleaning operation (years, periods, movements, and signatures) has been correctly committed to the main DataFrame.

In [69]:
# 1. Check for any remaining nulls in the entire DataFrame
null_report = df.isnull().sum()

# 2. Check for empty strings or 'nan' strings (Corrected logic)
# We apply map(str.lower) to each element to find hidden 'nan' texts
empty_strings = (df.astype(str) == '').sum()
hidden_nan_strings = df.astype(str).apply(lambda x: x.str.lower()).eq('nan').sum()

# 3. Verify specific column counts
print("--- FINAL NULL COUNT (Should be 0 everywhere) ---")
print(null_report)

print("\n--- HIDDEN EMPTY STRINGS ---")
print(empty_strings)

print("\n--- HIDDEN 'NAN' STRINGS ---")
print(hidden_nan_strings)

# 4. Final Row count verification
print(f"\nTotal final records: {len(df)} (Expected: 754)")

--- FINAL NULL COUNT (Should be 0 everywhere) ---
price                  0
artist                 0
title                  0
signed               153
condition              7
period                 0
movement               0
year_numeric           0
condition_grouped      0
is_signed              0
dtype: int64

--- HIDDEN EMPTY STRINGS ---
price                0
artist               0
title                0
signed               0
condition            0
period               0
movement             0
year_numeric         0
condition_grouped    0
is_signed            0
dtype: int64

--- HIDDEN 'NAN' STRINGS ---
price                0
artist               0
title                0
signed               0
condition            0
period               0
movement             0
year_numeric         0
condition_grouped    0
is_signed            0
dtype: int64

Total final records: 754 (Expected: 754)


## FINAL POLISHING - ACHIEVING ZERO NULL VARIANCE 
The final validation identified 7 residual nulls in the condition attribute and the redundant original signed column. To ensure a pristine dataset for Business Intelligence, the original signed column is removed, and the remaining condition gaps are filled with 'Not specified'.

In [70]:
# 1. Eliminate the redundant 'signed' column (The one with 153 nulls)
if 'signed' in df.columns:
    df.drop(columns=['signed'], inplace=True)

# 2. Fix the 7 remaining nulls in 'condition'
df['condition'] = df['condition'].fillna('Not specified')

# 3. Double check: ensure condition_grouped is also clean
df['condition_grouped'] = df['condition_grouped'].fillna('Good') # O el valor que prefieras por defecto

print("Final cleanup executed.")
print(f"Remaining nulls in condition: {df['condition'].isnull().sum()}")

Final cleanup executed.
Remaining nulls in condition: 0


Check one more time for a good quality data

In [71]:
# 1. THE ABSOLUTE ZERO CHECK (No nulls, no 'nan' strings)
nulls = df.isnull().sum().sum()
nans = df.astype(str).apply(lambda x: x.str.lower()).eq('nan').sum().sum()

# 2. DIMENSION CHECK
rows, cols = df.shape

# 3. LOGIC CHECK (Preview of our hard work)
print(f"--- FINAL AUDIT REPORT ---")
print(f"Total Records: {rows} | Total Columns: {cols}")
print(f"Total Nulls: {nulls}")
print(f"Total 'nan' Strings: {nans}")
print("-" * 30)

# Displaying a cross-section of the most important columns
display(df[['artist', 'price', 'year_numeric', 'is_signed', 'condition_grouped', 'period', 'movement']].head(10))

# Quick frequency check to see if categories look 'healthy'
print("\n--- QUICK CATEGORY HEALTH CHECK ---")
print(df['is_signed'].value_counts())
print("-" * 15)
print(df['period'].value_counts())

--- FINAL AUDIT REPORT ---
Total Records: 754 | Total Columns: 9
Total Nulls: 0
Total 'nan' Strings: 0
------------------------------


,artist,price,year_numeric,is_signed,condition_grouped,period,movement
0,Tommaso Ottieri,28500,2021,Yes,Excellent,Contemporary,Baroque
1,Pavel Tchelitchew,3000,1930,Yes,Other,Post-War,Surrealism
2,Leo Gabin,5000,2016,Yes,Excellent,Contemporary,Abstract
3,Matthias Dornfeld,5000,2010,Yes,Excellent,Contemporary,Abstract
4,Alexis Marguerite Teplin,2500,2001,Yes,Excellent,Contemporary,Abstract
5,Kenzo Okada,7575,1977,Yes,Excellent,Contemporary,Abstract Expressionism
6,Francesco Clemente,7550,2007,Yes,Excellent,Contemporary,Neo-Expressionism
7,Günther Förg,3550,1993,Yes,Good,Contemporary,Modernism
8,Dan Walsh,3075,2014,Yes,Excellent,Contemporary,Minimalism
9,Günther Förg,3550,1993,Yes,Excellent,Contemporary,Modernism



--- QUICK CATEGORY HEALTH CHECK ---
is_signed
Yes    601
No     153
Name: count, dtype: int64
---------------
period
Contemporary    415
Post-War        285
Modern           42
19th Century     12
Name: count, dtype: int64


## STEP X - DATASET FINALIZATION:
The data cleaning lifecycle is officially concluded. The resulting 'Golden Record' contains 754 observations with 100% data density across all dimensions. This clean CSV file serves as the definitive source for the upcoming exploratory data analysis and visualization phase.

In [72]:
# The definitive export
df.to_csv('Art_Dataset_Clean_Final.csv', index=False)

print("--------------------------------------------------")
print("✅ SUCCESS: 'Art_Dataset_Clean_Final.csv' is ready!")
print("--------------------------------------------------")

--------------------------------------------------
✅ SUCCESS: 'Art_Dataset_Clean_Final.csv' is ready!
--------------------------------------------------


In [73]:
# 1. Structural Audit
print("--- FINAL DATASET STRUCTURE (INFO) ---")
df.info()

--- FINAL DATASET STRUCTURE (INFO) ---
<class 'pandas.DataFrame'>
RangeIndex: 754 entries, 0 to 753
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   price              754 non-null    int64
 1   artist             754 non-null    str  
 2   title              754 non-null    str  
 3   condition          754 non-null    str  
 4   period             754 non-null    str  
 5   movement           754 non-null    str  
 6   year_numeric       754 non-null    int64
 7   condition_grouped  754 non-null    str  
 8   is_signed          754 non-null    str  
dtypes: int64(2), str(7)
memory usage: 53.1 KB


In [74]:
# 2. Statistical Audit (Numerical and Categorical)
print("\n--- FINAL STATISTICAL SUMMARY (DESCRIBE) ---")
# Usamos include='all' para ver tanto los números como las categorías más frecuentes
display(df.describe(include='all').T)


--- FINAL STATISTICAL SUMMARY (DESCRIBE) ---


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
price,754.0,NaN,NaN,NaN,6760.769231,12828.990543,595.0,1000.0,2500.0,7000.0,160000.0
artist,754,455,Russell Young,17,NaN,NaN,NaN,NaN,NaN,NaN,NaN
title,754,679,Untitled,35,NaN,NaN,NaN,NaN,NaN,NaN,NaN
condition,754,376,Excellent condition.,82,NaN,NaN,NaN,NaN,NaN,NaN,NaN
period,754,4,Contemporary,415,NaN,NaN,NaN,NaN,NaN,NaN,NaN
movement,754,35,Realism,178,NaN,NaN,NaN,NaN,NaN,NaN,NaN
year_numeric,754.0,NaN,NaN,NaN,1989.419098,26.284872,1850.0,1978.0,1990.0,2008.0,2050.0
condition_grouped,754,4,Good,300,NaN,NaN,NaN,NaN,NaN,NaN,NaN
is_signed,754,2,Yes,601,NaN,NaN,NaN,NaN,NaN,NaN,NaN
